### PR TITLE
fix(typescript): fix inlay hints mapping for large chunks of source code mapped verbatim to generated code.

### PR DESCRIPTION
**Pull request by piotrtomiak**
I've started working on support of inlay hints for Angular templates and I've noticed some issues.

### CODE REVIEW

File: packages/typescript/lib/node/proxyLanguageService.ts [hunk 1]
```diff
@@ -936,14 +936,29 @@ function provideInlayHints(language: Language<string>, provideInlayHints: ts.Lan
 		if (serviceScript) {
 			let start: number | undefined;
 			let end: number | undefined;
-			const map = language.maps.get(serviceScript.code, targetScript);
+			const map = language.maps.get(serviceScript.code, sourceScript);
```
Comment (piotrtomiak) [inline packages/typescript/lib/node/proxyLanguageService.ts:939]: First of all maps were taken for the wrong script. It should be a source script to have correct mappings from the template to generated code

File: packages/typescript/lib/node/proxyLanguageService.ts [hunk 2]
```diff
@@ -936,14 +936,29 @@ function provideInlayHints(language: Language<string>, provideInlayHints: ts.Lan
...
+				if (!isInlayHintsEnabled(mapping.data)) {
+					continue;
+				}
+				let mappingStart = mapping.sourceOffsets[0];
+				let genStart: number | undefined;
+				let genEnd: number | undefined;
+				if (mappingStart >= span.start && mappingStart <= span.start + span.length) {
+					genStart = mapping.generatedOffsets[0];
+					genEnd = mapping.generatedOffsets[mapping.generatedOffsets.length - 1]
+						+ (mapping.generatedLengths ?? mapping.lengths)[mapping.generatedOffsets.length - 1];
```
Comment (piotrtomiak) [inline packages/typescript/lib/node/proxyLanguageService.ts:950]: The end offset should include the whole length of the generated code.

File: packages/typescript/lib/node/proxyLanguageService.ts [hunk 3]
```diff
@@ -936,14 +936,29 @@ function provideInlayHints(language: Language<string>, provideInlayHints: ts.Lan
...
+				} else if (mappingStart < span.start && span.start < mappingStart + mapping.lengths[0]
+					&& mapping.sourceOffsets.length == 1
+					&& (!mapping.generatedLengths || mapping.generatedLengths[0] === mapping.lengths[0])
```
Comment (piotrtomiak) [inline packages/typescript/lib/node/proxyLanguageService.ts:953]: My Angular VirtualCode has large pieces of code copied verbatim from the original file. This condition ensures that it is correctly handled and not ignored.

**Comment by johnsoncodehk** [comment]
Since I had difficulty testing it, it took me some time to understand. Thank you for your PR!

Are you planning to make the same changes to getEncodedSemanticClassifications?

**Comment by piotrtomiak** [comment]
@johnsoncodehk - I haven't planned since we are not using these currently, so I would have hard time testing it. Thanks for merging the PR!

**Pull request merged**